### PR TITLE
Bayesian Personalized Ranking

### PR DESCRIPTION
**Pull request by benfred**
This adds a Bayesian Personalized Ranking model to implicit. This model
learns a matrix factorization embedding by minimizing the pairwise
ranking loss described in Bayesian Personalized Ranking from Implicit
Feedback(https://arxiv.org/pdf/1205.2618.pdf).

This commit adds a basic BPR MF model, using SGD to train the model.
A CUDA version of this model is also supplied, this model trains
about 10x faster on my system with identical results otherwise.

### CODE REVIEW

File: implicit/cuda/bpr.cu [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#include <cuda_runtime.h>
+#include <curand.h>
+#include <cublas_v2.h>
+
+
+#include "implicit/cuda/als.h"
+#include "implicit/cuda/utils.cuh"
+
+namespace implicit {
+__global__ void bpr_update_kernel(int samples, unsigned int * random_likes, unsigned int * random_dislikes,
```
Comment (maciejkula) [inline implicit/cuda/bpr.cu]: I really, really love this.

It's amazing to me that Hogwild works well even with such massive parallelism! How many parallel SGD updates does this actually do on a relatively good GPU?

Also, I _really_ need to get a GPU and get started on adding this to LightFM :)

Comment (benfred) [reply implicit/cuda/bpr.cu]: Thanks! 

On my GPU there are 3584 cuda cores - and each one is running a thread in parallel. 

I found CUDA programming relatively straightforward once I really wrapped my head around SIMT processing with warps of threads running the exact same instruction at the same time. I think that this CUDA code is even relatively easy to follow along with (though compared to lightfm it's a pretty simple model, I'm using SGD instead of adadelta, doesn't support item/user side features etc).

I just read your mixture paper btw - nice work! I really like the idea of treating users as a mixture of interests, and I'm going to forward it on to a couple people that I think will benefit from using that approach.

**Comment by eggie5** [comment]
This code offers orders of magnitude speedups compared to naive BPR SGD implementations. This is using what is described in the Hogwild! paper right? Parallel SGD updates using single CPU? Even better in this case you provided a GPU implementation of the gradient updates in CUDA code. 

What is the difference between this (the GPU implementation) and implemented BPR in an autodiff framework like Tensorflow or pytorch? Those are using the GPU but aren't doing the parallel  SGD updates right?

**Comment by benfred** [comment]
@eggie5 It is using a hogwild strategy for parallelization: we are not doing any locking when updating the factors for both the CPU and GPU variants.  This strategy is pretty common though and I doubt that it's responsible alone for any speedups you've seen.

What are you comparing this to for timing? On my system, the GPU  code is only about 10x faster than the CPU code - which is quite a bit less than the gains you were talking about on twitter.

I think the big difference with this code compared to PyTorch is in how it handles sparse updates.  When sampling a user/liked item/disliked item triple - this code only updates the factors for the sampled users and items. PyTorch doesn't support sparse operations like this as far I know. Last I checked, PyTorch only supported [sparse autograd operations for embeddings](https://github.com/pytorch/pytorch/issues/2389) - and sampling user/pos item/neg item like this didn't quite fit there. All told this means that PyTorch might be updating your entire user/item factors layers for each minibatch, even when only a small fraction of the  items/users have changed.  

I haven't actually tested this on PyTorch though, so all this is really just a conjecture =).

If you’re interested in running code on top of PyTorch - check out the Spotlight package, it’s pretty cool and includes more advanced loss functions like adaptive hinge loss that will probably lead to better results. The author of spotlight has also done some interesting work with bloom embeddings : https://github.com/maciejkula/spotlight/tree/master/examples/bloom_embeddings that might be of interest to you.

**Comment by maciejkula** [comment]
@benfred is right: PyTorch is not a great fit for relatively lightweight and very sparse models (like matrix factorization). This is for a combination of reasons:

1. As you mention, there are no sparse updates, and you need to update _all_ the embeddings.
2. Since you need to update all embeddings, you should really be using a GPU: it will do the work much faster.
3. Still, because there is little computation in the model itself (a simple dot product), you're at an awkward point of the CPU-GPU tradeoff curve where the CPU is too slow, but the GPU is severely underutilized, making kernel launches very costly relative to the core computation. The only way to get around this is to use extremely large minibatches, and this in turn reduces the model quality (in my experience).

To address these issues I have been working on my own autodifferentiation library which handles sparse gradients much more naturally: it's called [wyrm](https://github.com/maciejkula/wyrm). I've used  it to build a recommender package ([sbr](https://github.com/maciejkula/sbr-rs)). I am quite happy with the end result (although it is definitely not as fast as @benfred's specialized solution). It currently has no Python bindings, but you can use it from Go if that's something you're interested in ([sbr](https://github.com/maciejkula/sbr-go)).

**Comment by ita9naiwa** [comment]
check this out too[vectorflow](https://github.com/Netflix/vectorflow).

This is MF or neural network library for sparse dataset maintained by netflix(its not very active now but...) worth check it.
